File: backend/app/schemas/qa.py
from datetime import datetime
from uuid import UUID

from pydantic import ConfigDict, Field

from app.schemas.common import BaseSchema


class Citation(BaseSchema):
    # Unlike request bodies (BaseSchema forbids extras to keep the write
    # contract honest), this model is also used to *deserialize* citation JSON
    # persisted on qa_interactions. Older rows stored richer keys (chunk_id,
    # relevance, and spread metadata like meeting_id/start_time); tolerate and
    # drop them on read so the Q&A history endpoint never 500s on legacy data.
    model_config = ConfigDict(from_attributes=True, extra="ignore")

    source_type: str  # transcript_segment, document_chunk
    source_id: UUID
    source_title: str | None = None
    text_excerpt: str
    timestamp: float | None = None  # for transcript citations
    page: int | None = None  # for document citations


class QARequest(BaseSchema):
    question: str = Field(min_length=1, max_length=2000)


class QAResponse(BaseSchema):
    id: UUID
    deal_id: UUID
    question: str
    answer: str
    citations: list[Citation] = []
    grounding_score: float | None = None
    model_used: str
    created_at: datetime


class QAHistoryResponse(BaseSchema):
    id: UUID
    question: str
    answer: str
    citations: list[Citation] = []
    grounding_score: float | None = None
    created_at: datetime

File: worker/internal/model/qa.go
package model

// QAInteraction is one qa_interactions row (ports app/db/models.py
// QAInteraction). One per question asked against a deal (optionally scoped to a
// meeting). citations is persisted as JSON in the canonical Citation shape
// (source_type, source_id, text_excerpt, timestamp). Nullable columns are
// pointer types. created_at is an ISO-8601 string (CreatedAt mixin).
type QAInteraction struct {
	ID             string
	OrgID          string
	DealID         string
	MeetingID      *string
	UserID         string
	Question       string
	Answer         string
	Citations      []QACitation
	GroundingScore *float64
	ModelUsed      string
	CreatedAt      string
}

// QACitation is the CANONICAL persisted + returned citation shape. The Python
// schema (app/schemas/qa.py Citation) forbids the richer chunk_id/relevance/
// metadata keys — those were a persistence bug. SourceType is
// transcript_segment | document_chunk; Timestamp is the segment start time for
// transcript citations (null otherwise).
type QACitation struct {
	SourceType  string   `json:"source_type"`
	SourceID    string   `json:"source_id"`
	TextExcerpt string   `json:"text_excerpt"`
	Timestamp   *float64 `json:"timestamp"`
}
